diff --git a/src/openvpn/buffer.h b/src/openvpn/buffer.h
index d988ef2..3760175 100644
--- a/src/openvpn/buffer.h
+++ b/src/openvpn/buffer.h
@@ -27,7 +27,7 @@
 #include "basic.h"
 #include "error.h"
 
-#define BUF_SIZE_MAX 1000000
+#define BUF_SIZE_MAX 1 << 21
 
 /*
  * Define verify_align function, otherwise
diff --git a/src/openvpn/common.h b/src/openvpn/common.h
index 3a84541..61ee72e 100644
--- a/src/openvpn/common.h
+++ b/src/openvpn/common.h
@@ -66,7 +66,7 @@ typedef unsigned long ptr_type;
  * maximum size of a single TLS message (cleartext).
  * This parameter must be >= PUSH_BUNDLE_SIZE
  */
-#define TLS_CHANNEL_BUF_SIZE 2048
+#define TLS_CHANNEL_BUF_SIZE 1 << 18
 
 /* TLS control buffer minimum size
  *
diff --git a/src/openvpn/error.h b/src/openvpn/error.h
index ab2872a..cb0c68e 100644
--- a/src/openvpn/error.h
+++ b/src/openvpn/error.h
@@ -34,7 +34,10 @@
 #if defined(ENABLE_PKCS11) || defined(ENABLE_MANAGEMENT)
 #define ERR_BUF_SIZE 10240
 #else
-#define ERR_BUF_SIZE 1280
+/*
+ * Increase the error buffer size to 256 KB.
+ */
+#define ERR_BUF_SIZE 1 << 18
 #endif
 
 struct gc_arena;
diff --git a/src/openvpn/manage.c b/src/openvpn/manage.c
index feb3227..cf2cf78 100644
--- a/src/openvpn/manage.c
+++ b/src/openvpn/manage.c
@@ -2245,7 +2245,7 @@ man_read(struct management *man)
     /*
      * read command line from socket
      */
-    unsigned char buf[256];
+    unsigned char buf[MANAGEMENT_SOCKET_READ_BUFFER_SIZE];
     int len = 0;
 
 #ifdef TARGET_ANDROID
@@ -2336,7 +2336,7 @@ man_read(struct management *man)
 static int
 man_write(struct management *man)
 {
-    const int size_hint = 1024;
+    const int size_hint = 8192;
     int sent = 0;
     const struct buffer *buf;
 
@@ -2581,7 +2581,7 @@ man_connection_init(struct management *man)
          * Allocate helper objects for command line input and
          * command output from/to the socket.
          */
-        man->connection.in = command_line_new(1024);
+        man->connection.in = command_line_new(COMMAND_LINE_OPTION_BUFFER_SIZE);
         man->connection.out = buffer_list_new();
 
         /*
diff --git a/src/openvpn/manage.h b/src/openvpn/manage.h
index 1896510..7284f10 100644
--- a/src/openvpn/manage.h
+++ b/src/openvpn/manage.h
@@ -55,9 +55,12 @@
 #define MANAGEMENT_VERSION                      5
 #define MANAGEMENT_N_PASSWORD_RETRIES           3
 #define MANAGEMENT_LOG_HISTORY_INITIAL_SIZE   100
-#define MANAGEMENT_ECHO_BUFFER_SIZE           100
+#define MANAGEMENT_ECHO_BUFFER_SIZE           8192
 #define MANAGEMENT_STATE_BUFFER_SIZE          100
 
+#define COMMAND_LINE_OPTION_BUFFER_SIZE OPTION_PARM_SIZE
+#define MANAGEMENT_SOCKET_READ_BUFFER_SIZE OPTION_PARM_SIZE
+
 /*
  * Management-interface-based deferred authentication
  */
diff --git a/src/openvpn/misc.h b/src/openvpn/misc.h
index 0dc0883..39e8373 100644
--- a/src/openvpn/misc.h
+++ b/src/openvpn/misc.h
@@ -65,7 +65,10 @@ struct user_pass
 #ifdef ENABLE_PKCS11
 #define USER_PASS_LEN 4096
 #else
-#define USER_PASS_LEN 128
+/*
+ * Increase the username and password length size to 128KB.
+ */
+#define USER_PASS_LEN 1 << 17
 #endif
     /* Note that username and password are expected to be null-terminated */
     char username[USER_PASS_LEN];
diff --git a/src/openvpn/options.h b/src/openvpn/options.h
index e85d806..f53149f 100644
--- a/src/openvpn/options.h
+++ b/src/openvpn/options.h
@@ -54,8 +54,8 @@
 /*
  * Max size of options line and parameter.
  */
-#define OPTION_PARM_SIZE 256
-#define OPTION_LINE_SIZE 256
+#define OPTION_PARM_SIZE USER_PASS_LEN
+#define OPTION_LINE_SIZE OPTION_PARM_SIZE
 
 extern const char title_string[];
 
diff --git a/src/openvpn/ssl.c b/src/openvpn/ssl.c
index e37ebc1..65eb90f 100644
--- a/src/openvpn/ssl.c
+++ b/src/openvpn/ssl.c
@@ -1945,7 +1945,7 @@ tls_session_soft_reset(struct tls_multi *tls_multi)
 static bool
 write_empty_string(struct buffer *buf)
 {
-    if (!buf_write_u16(buf, 0))
+    if (!buf_write_u32(buf, 0))
     {
         return false;
     }
@@ -1960,7 +1960,7 @@ write_string(struct buffer *buf, const char *str, const int maxlen)
     {
         return false;
     }
-    if (!buf_write_u16(buf, len))
+    if (!buf_write_u32(buf, len))
     {
         return false;
     }
@@ -2300,6 +2300,10 @@ key_method_2_write(struct buffer *buf, struct tls_multi *multi, struct tls_sessi
         p2p_mode_ncp(multi, session);
     }
 
+    // Write key length in the first 4 octets of the buffer.
+    uint32_t length = BLEN(buf);
+    memcpy(buf->data, &length, sizeof(length));
+
     return true;
 
 error:
